### PR TITLE
Added before_dependent_destroy

### PR DESCRIPTION
**Pull request by AndyObtiva**
As per my comment on this issue:
https://github.com/rails/rails/issues/670#issuecomment-111306103

I realize there are existing alternatives, such as prepend: true, but please hear me out and put yourself in the shoe of a Rails framework consumer who sees Rails as a black box during business development, and rightly so as that is part of what made Rails so successful.

On a recent Android/Rails project, we got surprised by the default sequencing of events for dependent: :destroy vs a before_destroy added below it in an activerecord. From a Rails framework consumer point of view, the thought process assumes dependent: :destroy behavior to be part of #destroy. So, we expected before_destroy to happen before any kind of destruction on the active record and all its dependent destroy active records. That wasn't the case because from a Rails framework developer point of view, dependent: :destroy relies on before_destroy hooks in its implementation (something a consumer would not know intuitively without reading and digging). One hour and a half of troubleshooting later, I discovered the cause of our project problem. 

Still, I think we can save many Rails developers trouble in the future by having this adhere to the principle of least surprise (aka, before_destroy happens before any kind of destruction on the main active record or dependent ones). Given there is already a prepend:true option and we don't want to break older Rails applications, one friendly alternative is to simply support a more intuitive before_dependent_destroy, and that is what i implemented in this Pull Request and added documentation for. It would help make things absolutely clear with regards to the ordering of events when a developer learns of it alongside before_destroy.

This is a first cut implementation of before_dependent_destroy that I did last night test-first (activerecord/test/cases/callbacks_test.rb). Your feedback is greatly appreciated. I just noticed I am missing tests for after_dependent_destroy and around_dependent_destroy, so I'm adding soon (please bare with me).

Thanks. I've been a happy camper with Rails and appreciate the Rails folks continually striving to meet developer needs. I hope this solution pleases the happiness standards you conceive of for the Rails community.

Regards,

Andy Maleh

### CODE REVIEW

File: activerecord/test/cases/callbacks_test.rb [hunk 1]
```diff
@@ -78,6 +78,42 @@ def cancel
...
+end
+
+class RelatedDeveloper < ActiveRecord::Base
+  self.table_name = 'developers'
+end
+
+class RelatedDeveloperProject < ActiveRecord::Base
+  self.table_name = 'developers_projects'
+
+  belongs_to :developer, :class => RelatedProject, :foreign_key => 'developer_id'
```
Comment (rafaelsales) [inline activerecord/test/cases/callbacks_test.rb]: Since no backwards compatibility is planned for this, we can use Ruby >= 1.9 syntax for hashes

Comment (AndyObtiva) [inline activerecord/test/cases/callbacks_test.rb]: Thanks. That's great to know. Indeed, I wasn't sure whether the new hash syntax would be allowed. Forgot Ruby 1.9 support has been dropped from newer Rails releases. I'll use the new hash syntax in my code for my upcoming revisions on this Pull Request.

**Comment by sgrif** [comment]
Closing due to inactivity.

**Comment by RochesterinNYC** [comment]
@AndyObtiva @sgrif Anyone opposed if I pick this up? Would like to see/use this in Rails 5...

**Comment by AndyObtiva** [comment]
No worries. I'd say go ahead. I've been wanting to pick it up again for a while and hadn't had a chance. Let me know if you'd like to pair on it. I might be available and can always appreciate collaborating with someone I haven't worked with before.

**Comment by AndyObtiva** [comment]
@RochesterinNYC it was such a no-brainer, I just revised the commit for Rails 5 latest master and submitted here: https://github.com/rails/rails/pull/22520 (I'd be happy to accept additional suggestions though since you raised this up and motivated me to redo for Rails 5)

**Pull request closed without merging**